shell/exec.go

```diff
@@ -102,8 +103,24 @@ func (s *MenderShell) pipeStdout() {
 		raw := make([]byte, 255)
 		n, err := sr.Read(raw)
 		if err != nil {
-			log.Errorf("error reading stdout: '%s'; restart is needed.", err)
-			break
+			log.Errorf("error reading stdout: %s", err)
+			//
+			msg := &ws.ProtoMsg{
+				Header: ws.ProtoHdr{
+					Proto:     ws.ProtoTypeShell,
+					MsgType:   wsshell.MessageTypeStopShell,
+					SessionID: s.sessionId,
+					Properties: map[string]interface{}{
+						"status": wsshell.NormalMessage,
+					},
+				},
+				Body: []byte(err.Error()),
+			}
+			err = connectionmanager.Write(ws.ProtoTypeShell, msg)
+			if err != nil {
+				log.Debugf("error on write: %s", err.Error())
+			}
+			return
 		} else if !s.IsRunning() {
```

Comment:
What about this case (and before calling Read)? There is a potential race where the terminal can terminate outside blocking on `sr.Read`.

Reply:
If it terminates by other means (e.g. shutdown of the daemon), we manage it.
I believe the only case where we did not manage a clean shutdown is the termination of the process, which we detect with the inability to read its stdout. I may be wrong, though.

Reply:
Yeah, naturally the file descriptors will close and return `io.EOF` (or `io.ErrUnexpectedEOF`?). Disregard...